modules/margin/src/main/java/com/opengamma/sdk/margin/InvokerMarginClient.java

```diff
@@ -177,6 +180,67 @@ public MarginCalcResult calculate(Ccp ccp, MarginCalcRequest request) {
   }
 
   @Override
+  public MarginWhatIfCalcResult calculateWhatIf(Ccp ccp, MarginCalcRequest request, List<PortfolioDataFile> deltaFiles) {
+    String baseCalcId = createCalculation(ccp, request);
+    MarginCalcRequest secondRequest = MarginCalcRequest.builder()
+        .portfolioData(deltaFiles)
+        .calculationCurrency(request.getCalculationCurrency().orElse(""))
+        .reportingCurrency(request.getReportingCurrency())
+        .valuationDate(request.getValuationDate())
+        .build();
+    String deltaCalcId = createCalculation(ccp, secondRequest);
+
+    MarginCalcResult baseResult = getCalculation(ccp, baseCalcId);
+    MarginCalcResult deltaResult = getCalculation(ccp, deltaCalcId);
+    while (MarginCalcResultStatus.PENDING.equals(baseResult.getStatus()) || MarginCalcResultStatus.PENDING.equals(
+        deltaResult.getStatus())) {
+      try {
+        Thread.sleep(POLL_WAIT);
+      } catch (InterruptedException ex) {
+        throw new RuntimeException(ex);
+      }
+      if (MarginCalcResultStatus.PENDING.equals(baseResult.getStatus())) {
+        baseResult = getCalculation(ccp, baseCalcId);
+      }
+      if (MarginCalcResultStatus.PENDING.equals(deltaResult.getStatus())) {
+        deltaResult = getCalculation(ccp, deltaCalcId);
+      }
+    }
+    // cleanup server state quietly
+    try {
+      deleteCalculation(ccp, baseCalcId);
+      deleteCalculation(ccp, deltaCalcId);
+    } catch (RuntimeException ex) {
+      // ignore
+    }
+
+    MarginSummary baseSummary = baseResult.getMargin().orElseThrow(IllegalStateException::new);
```

Comment:
This logic should move to a package scoped factory method on `MarginWhatIfCalcResult` that takes the two separate results.